node/src/components/linear_chain_sync/operations.rs

```diff
@@ -451,11 +443,17 @@ where
     // Fetch the trusted header
     let trusted_header = fetch_and_store_block_header(effect_builder, trusted_hash).await?;
 
-    let mut trusted_validator_weights =
-        get_trusted_validator_weights(effect_builder, &chainspec, &trusted_header).await?;
+    let mut maybe_trusted_switch_block =
+        maybe_get_trusted_switch_block(effect_builder, &chainspec, &trusted_header).await?;
+
+    let mut trusted_validator_weights = maybe_trusted_switch_block.as_ref().map_or_else(
+        || get_genesis_validators(&chainspec),
+        |switch_block| switch_block.next_era_validator_weights().unwrap().clone(),
```

Comment:
Is this correct? What if, while walking back to the last switch block, we'll cross an emergency restart boundary? I.e., we're unlucky and the last emergency restart was right when the era with the trusted block started?

Reply:
I think fast sync simply doesn't support joining in the era after an emergency restart yet, and it's not yet decided how that support will be implemented in the future. (@xcthulhu: please correct me if I'm wrong.)

Reply:
And the code before this PR had the same behavior, if I'm reading it correctly.

Reply:
I think this could be resolved by a simple call to `effect_builder.get_era_validators(trusted_header.era_id())` - it should handle all the logic correctly (though this point is _before_ we finish syncing, if I understand correctly, and I'm not sure how well that plays with the assumptions `get_era_validators` makes :thinking: ).

Reply:
I see. So I'd mark it as a TODO for now, and we'll have to figure out how to handle such a case properly.

Reply:
> I think this could be resolved by a simple call to `effect_builder.get_era_validators(trusted_header.era_id())` - it should handle all the logic correctly (though this point is before we finish syncing, if I understand correctly, and I'm not sure how well that plays with the assumptions `get_era_validators` makes thinking ).

Maybe we should make `get_era_validators` a private function for the `era_supervisor`, rather than a method on the effect builder?  We can't call it in general from the joiner reactor, and I don't think anyone else is using it.

The assumption that we will only join after the first era after an emergency restart is probably too much like you say, we'll figure something out.